### PR TITLE
Add api module with `download_to` method

### PR DESCRIPTION
**Pull request by theory**
Use the `ureq` crate for simple synchronous requests, but keep it encapsulated so we can replace it later if need be. Create an agent that supports proxies and require HTTPS for all requests. Use it in the `download_to` method, but if the API base URL has a `file` schema, skip the ureq agent and copy the file directly. This will be useful for working with `rsync`d mirrors.

Add variants to the BuildError enum for IO, URL parsing and formatting, and HTTP request errors from ureq. Box ureq errors because otherwise they're far to big, and [clippy complains] about the error size. Make a custom File error to improve IO errors to include the action attempted and relevant file name or names.

Use `httpmock` to test the handling of `http:` URLs, and `sha2` & `tempfile` to test `file:` URLs.

While at it, exclude `Cargo.lock` from `git diff`; it's just too verbose and distracting. Also use string comparison for some error tests in `src/tests.rs`.

  [clippy complains]: https://rust-lang.github.io/rust-clippy/master/index.html#result_large_errclippyClick

### CODE REVIEW

File: src/api/mod.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                    let dst = dir.as_ref().join(filename);
+
+                    if url.scheme() == "file" {
+                        // Just copy the file.
+                        let src = match url.to_file_path() {
+                            Err(_) => Err(BuildError::NoUrlFile(url.clone()))?,
+                            Ok(s) => s,
+                        };
+
+                        // Eschew std::fs::copy for better error messages.
```
Comment (vrmiguel) [inline src/api/mod.rs:70]: There's a crate called [fs-err](https://docs.rs/fs-err/latest/fs_err/) that is essentially a wrapper over `std::fs` but with reasonable error messages! You might find it useful.

It'd be something like this:

```rust
use fs_err as fs;

// ..

fs::copy(src, dst)?;
```

Comment (theory) [reply src/api/mod.rs:70]: That's handy, but it sadly doesn't have a wrapper for [io::copy](https://doc.rust-lang.org/stable/std/io/fn.copy.html), only for [fs::copy](https://doc.rust-lang.org/stable/std/fs/fn.copy.html).

Comment (theory) [reply src/api/mod.rs:70]: Also I just tried it, and alas it is not a perfect drop-in replacement for `fs`. `fs::File::open`, for example, has this signature:

```rs
pub fn open<P: AsRef<Path>>(path: P) -> io::Result<File>
```

While fs-err has:

```rs
pub fn open<P>(path: P) -> Result<Self, io::Error>
    where
        P: Into<PathBuf>,
    {
```

I assume it mostly works, but not when the input is _defined_ as `AsRef<Path>`, as in the `unpack` method I just added in #7.

File: src/error/mod.rs [hunk 2]
```diff
@@ -11,4 +12,30 @@ pub enum BuildError {
     /// Unknown pipeline error.
     #[error("unknown build pipeline `{0}`")]
     UnknownPipeline(String),
+
+    /// IO rror.
```
Comment (vrmiguel) [inline src/error/mod.rs:16]: e

**Review by vrmiguel — APPROVED** [review]

File: src/error/mod.rs (hunk 2 above)
Comment (theory) [reply src/error/mod.rs]: Bah, fixed in b0f261a, thanks.

**Pull request merged**